src/main/java/com/amazon/opendistroforelasticsearch/knn/index/util/KNNLibrary.java

```diff
@@ -0,0 +1,214 @@
+     * are common across different underlying libraries.
+     */
+    abstract class NativeLibrary implements KNNLibrary {
+        protected Map<String, KNNMethod> methods;
+        private Map<SpaceType, Function<Float, Float>> scoreTranslation;
+        private String latestLibraryBuildVersion;
+        private String latestLibraryVersion;
+        private String extension;
+
+        /**
+         * Constructor for NativeLibrary
+         *
+         * @param methods map of methods the native library supports
+         * @param scoreTranslation Map of translation of space type to scores returned by the library
+         * @param latestLibraryBuildVersion String representation of latest build version of the library
+         * @param latestLibraryVersion String representation of latest version of the library
+         * @param extension String representing the extension that library files should use
+         */
+        public NativeLibrary(Map<String, KNNMethod> methods, Map<SpaceType, Function<Float, Float>> scoreTranslation,
+                             String latestLibraryBuildVersion, String latestLibraryVersion, String extension)
+        {
+            this.methods = methods;
+            this.scoreTranslation = scoreTranslation;
+            this.latestLibraryBuildVersion = latestLibraryBuildVersion;
+            this.latestLibraryVersion = latestLibraryVersion;
+            this.extension = extension;
+        }
+
+        @Override
+        public String getLatestBuildVersion() {
+            return this.latestLibraryBuildVersion;
+        }
+
+        @Override
+        public String getLatestLibVersion() {
+            return this.latestLibraryVersion;
+        }
+
+        @Override
+        public String getExtension() {
+            return this.extension;
+        }
+
+        @Override
+        public String getCompoundExtension() {
+            return getExtension() + "c";
+        }
+
+        @Override
+        public KNNMethod getMethod(String methodName) {
+            if (!methods.containsKey(methodName)) {
+                throw new IllegalArgumentException("Invalid method name: " + methodName);
+            }
+            return methods.get(methodName);
+        }
+
+        @Override
+        public float score(float rawScore, SpaceType spaceType) {
+            if (this.scoreTranslation.containsKey(spaceType)) {
+                return this.scoreTranslation.get(spaceType).apply(rawScore);
+            }
+
+            return spaceType.scoreTranslation(rawScore);
+        }
+
+        @Override
+        public void validateMethod(KNNMethodContext knnMethodContext) {
+            String methodName = knnMethodContext.getMethodComponent().getName();
+            if (!methods.containsKey(methodName)) {
+                throw new ValidationException();
+            }
+
+            KNNMethod knnMethod = methods.get(methodName);
+            knnMethod.validate(knnMethodContext);
```

Comment:
if it doesn't matter whether it is ValidationException or IllegalArgumentException, 
```suggestion
            getMethod(methodName).validate(knnMethodContext);
```
else you can catch and throw accordingly. IMO, it doesn't matter.